src/endpoint/s3/ops/s3_post_object_select.js

```diff
@@ -0,0 +1,92 @@
+const { S3SelectStream } = require('../../../util/s3select');
+const nb_native = require('../../../util/nb_native');
+const stream_utils = require('../../../util/stream_utils');
+
+/**
+ * https://docs.aws.amazon.com/AmazonS3/latest/API/API_SelectObjectContent.html
+ */
+async function post_object_select(req, res) {
+
+    if (!nb_native().S3Select) {
+        throw new S3Error(S3Error.S3SelectNotCompiled);
+    }
+
+    req.object_sdk.setup_abort_controller(req, res);
+    const agent_header = req.headers['user-agent'];
+    const noobaa_trigger_agent = agent_header && agent_header.includes('exec-env/NOOBAA_FUNCTION');
+    const encryption = s3_utils.parse_encryption(req);
+
+    const md_params = {
+        bucket: req.params.bucket,
+        key: req.params.key,
+        version_id: req.query.versionId,
+        encryption,
+    };
+    const object_md = await req.object_sdk.read_object_md(md_params);
+
+    const params = {
+        object_md,
+        obj_id: object_md.obj_id,
+        bucket: req.params.bucket,
+        key: req.params.key,
+        content_type: object_md.content_type,
+        noobaa_trigger_agent,
+        encryption,
+    };
+
+    //prepare s3select stream
+    const http_req_select_params = req.body.SelectObjectContentRequest;
+    const input_serialization = http_req_select_params.InputSerialization[0];
+    let input_format = null;
+    if ('CSV' in input_serialization) {
+        input_format = 'CSV';
+    } else if ('JSON' in input_serialization) {
+        input_format = 'JSON';
+    } else {
+        throw new S3Error(S3Error.MissingInputSerialization);
+    }
+    const select_args = {
+        query: http_req_select_params.Expression[0],
+        input_format: input_format,
+        input_serialization_format: http_req_select_params.InputSerialization[0][input_format][0],
+        records_header_buf: S3SelectStream.records_message_headers
+    };
+    const s3select = new S3SelectStream(select_args);
+    dbg.log3("select_args = ", select_args);
+
+    //pipe s3select result into http result
+    stream_utils.pipeline([s3select, res], true);
+
+    //send s3select pipe to read_object_stream.
+    //in some cases (currently nsfs) it will pipe object stream into our pipe (s3select)
+    const read_stream = await req.object_sdk.read_object_stream(params, s3select);
+    if (read_stream) {
+        // if read_stream supports closing, then we handle abort cases such as http disconnection
+        // by calling the close method to stop it from buffering more data which will go to waste.
+        if (read_stream.close) {
+            req.object_sdk.add_abort_handler(() => read_stream.close());
+        }
+        read_stream.on('error', err => {
+            dbg.error('read stream error:', err, req.path);
+            res.destroy(err);
+        });
+        //in other cases, we need to pipe the read stream ourselves
+        stream_utils.pipeline([read_stream, s3select], true);
```

Comment:
@romayalon - We need to check about the last bool arg `reuse_last_stream` with you.
As the code is now, the last stream in this pipeline call is the s3select stream, which was already pipelined into res above, so it might not need to be resumed, but I'm not sure after reading the comment about it in stream_utils.pipeline.

Reply:
if s3select stream is not needed anymore we need to call resume(), so the true argument should not be passed to the function.

Reply:
i am still confused - what does it mean "not needed anymore"?

Reply:
the s3select  stream won't be used after that line